### DISCUSSION TITLE
Authenticating with Seqr for API access without browser

### DISCUSSION THREAD

**Discussion by tommyli** [broadinstitute/seqr#2037] [Q&A] [answered]
We managed to integrate our instance of Seqr with Okta, MCRIs identity provider using OAuth2 + OpenID.  [python-social-auth](https://python-social-auth.readthedocs.io/en/latest/index.html) was a good choice as it allowed us to customise the auth flows to our needs using [pipelines](https://python-social-auth.readthedocs.io/en/latest/pipeline.html).  Thanks very much for that.

We'd also like to know how Broad access Seqr APIs using non browser based clients?  E.g. using curl or something.  This use case allows potential for further automation, e.g. "bulk edit individuals", "edit datasets", bulk load HPO terms etc.

We implemented something using [Django OAuth Toolkit (Resource Server)](https://django-oauth-toolkit.readthedocs.io/en/latest/resource_server.html#setup-the-resource-server) which allows clients to authorise themselves with Seqr by providing Bearer access tokens in the HTTP header.  Our [implementation](https://github.com/bioinfomethods/seqr/tree/feat-mcri-okta-security) is very rough and fails CI but should illustrate the changes required.

Any thoughts?

**Comment by hanars** [comment, marked as answer]
The short answer is we really don't access seqr via API. We have a non-technical project management team that is responsible for maintaining seqr data so we have really optimized for that. 

If we do have things we want to run automatically against seqr out current approach is to write manage commands to do them and then add a cron job to run directly in the seqr pod to execute them. That way, the job is running in the authenticated environment and does not require user credentials. 

We also do have one internal script we periodically run against seqr APIs, and for that we have the user log in to seqr and then copy their session_id cookie and pass it to their script. This doesn't work for CI probably, as these tokens expire within 24 hours, but works for ad-hoc jobs
```
 with requests.Session() as s:
        s.cookies.set_cookie(requests.cookies.create_cookie('sessionid', session_id))
```

**Reply by tommyli** [reply]
Thanks for that insight, that clarifies things.

Would you like me to raise a PR (not to be accepted) as an example implementation for future reference?  I've cleaned up the implementation and now passes CI.

**Reply by hanars** [reply]
yeah that'd be cool thanks! Maybe put a link to the PR here so in the future people can find it?

**Reply by tommyli** [reply]
Raised #2041.

**Comment by illusional** [comment]
Hey @hanars, we're finding we're very interested in programmatic access to seqr - particularly for the automated interpretation pipeline work (eg: being able to provide direct links to a family search in seqr), and automating various aspects of the loda (eg: updating ES indexes).

Is programmatic access a change you're interested in being upstreamed - or do you have thoughts on ways this could be incorporated into seqr without impacting security requirements? Just want to get your thoughts on this before we go further down this rabbit hole.

**Comment by hanars** [comment]
So we are not opposed on principle to upstreaming programmatic access if it is securely done, but there are a lot of security implications to these changes and I would need to run them past our compliance certification team before we could accept them. The earlier we start that process the more likely such changes are to be accepted.  

Given these implications, we also will push back on changes if we think there is a different, equally effective solution that could be implemented without changing seqr's access. For example, for providing direct links to a family search in seqr it strikes me that the most secure way to do that would be to update seqr to support search URLs where the search is in the URL itself. Then those links could be generated outside of seqr and when an authenticated user accesses that link in the browser it would display the desired search. This is a much lower security risk than allowing a programmatic user to generate links, so we probably would not upstream programmatic access for this use case.

Updating ES indices automatically is something we have been thinking about doing for a long time, but we do not currently have the bandwidth on our end to implement it so we would be eager to accept solutions for that. However, I think we would prefer to use a cron job as I described above rather than  programmatic access. But if there is a really compelling reason to use programmatic access instead of cron I am open to hearing them. 

TL;DR we are not opposed on principle to incorporating programmatic access to seqr but we would prefer to explore other solutions first. If you have a particular use case you are planning to implement I encourage you to reach out to me and I can let you know before you start implementation if we would be interested in upstreaming said changes and what if any implementation details we would require to o so.

**Reply by illusional** [reply]
Thanks @hanars! My initial description was not very compelling, so I thought I'd clarify and ask for your feedback on my initial implementation. We currently have the following use cases in mind, which could all be solved elegantly through API access that’s granted to service accounts:

1. As part of the Broad / Microsoft collaboration, we’re working on an automated analysis pipeline (https://github.com/populationgenomics/automated-interpretation-pipeline) which will flag variants. We’d like to link those variants directly to their corresponding page in seqr, so curators can perform an in-depth analysis through seqr. We expect that this is also functionality that the Broad working group would be very interested in. We currently can’t generate these links directly, as we’d need to know the internal GUIDs for the family and project. Instead, we’d like to work together on an API to generate such links in a way that won’t depend on internal representation details.
2. Updating ES indexes automatically from the loading pipeline, as well as creating new datasets with associated metadata in an automated fashion.
3. Running integration tests that would perform a well curated set of variant searches on staging / production data to check that the result matches a “golden” result. This could catch a number of issues that aren’t easily covered by unit tests, e.g. subtle changes in the way that annotations are added.

We’d really appreciate your feedback on this proof of concept.  Happy to throw it away or make major structural changes if you have feedback that would allow for a better chance of upstreaming this change:

Summary of PR: https://github.com/populationgenomics/seqr/pull/130

> Introduces a `service_account_access` decorator that allows access by authorized service-accounts to specific endpoints.
>
> - Ensuring that the account that is authenticating belongs to the "service_account_access" group (configurable).
> - Nb: normal users cannot access this route programmatically or through the browser
> - [Open question]: Should we restrict accounts with service-account-access from accessing regular routes (ie: add not has_service_account_access to login_required check).
> - Disabling CSRF for routes with this decorator.
> - Enabling and requiring Bearer token auth for these routes.
>     - There is a `GoogleBearerAuth` implementation that validates a Google created JWT with the audience matching `SOCIAL_AUTH_GOOGLE_OAUTH2_CLIENT_ID`.
>     - This also restricts you to accessing only via a service account as google does not let you create a personal identity-token for an audience (ie: `ERROR: Invalid account Type for --audiences. Requires valid service account`). However, for service accounts such a token can be generated very easily using `gcloud auth print-identity-token`.

**Reply by hanars** [reply]
So for a little more context, as part of our FISMA certification it is a lot of work to make changes to key security components of seqr like authentication and authorization. It requires us to go through a review process and update a lot of documentation. While it is of course possible to make these changes, it is enough additional work that there really needs to be a strong defensible reason why we can not achieve the same results through an already approved security mechanism. 

Given the difficulty of updating and maintaining alternative modes of authentication to seqr, I don't think I can upstream any of these change to the main seqr repo if they include any changes to authentication. I of course invite you to maintain your own fork with such changes, as theres no requirement that forks of seqr conform to our internal security standards. I also would be happy to upstream these changes if they could be done using the current authentication mechanism.  Here are detailed responses for the use cases you outline and why I don't feel that they justify a new mode of authentication:

> 1. As part of the Broad / Microsoft collaboration, we’re working on an automated analysis pipeline (https://github.com/populationgenomics/automated-interpretation-pipeline) which will flag variants. We’d like to link those variants directly to their corresponding page in seqr, so curators can perform an in-depth analysis through seqr. We expect that this is also functionality that the Broad working group would be very interested in. We currently can’t generate these links directly, as we’d need to know the internal GUIDs for the family and project. Instead, we’d like to work together on an API to generate such links in a way that won’t depend on internal representation details.

I do understand the difficulty of mapping to internal GUIDs, but I still don't think an API is needed for these changes. GUIDs are stable so supporting a download from seqr of the necessary sample ID to family/ project GUID mapping should be a sufficient solution to the problem. Building out this download should not be too difficult. You would need to manually re-download and replace the file if you were running the analysis pipeline on a new set of samples, but I think that should not be too onerous.  It's also worth noting that seqr does not require that sample IDs are unique across projects, so having a file that can be manually tweaked actually has some advantages. For example, there are frequently samples for which we have both WES and WGS data, so maintaining separate mapping files grouped by sample type might be advantageous

> 2. Updating ES indexes automatically from the loading pipeline

As I have noted several times on this thread, this can be done using a cron job. We already have many cron jobs running in our seqr deployment and using one for this would be consistant and secure, and I don't see any significant advantages to an API call over a cron job. 

> as well as creating new datasets with associated metadata in an automated fashion.

I don't actually know what this means so I can't fully respond here

> 3. Running integration tests that would perform a well curated set of variant searches on staging / production data to check that the result matches a “golden” result. This could catch a number of issues that aren’t easily covered by unit tests, e.g. subtle changes in the way that annotations are added.

I do understand the importance of robust testing, but I don't see why added API access is necessary for this. You can easily set up a seqr installation that uses simple username/ password authentication (the default docker installation works this way) and then you can set up a python requests session, authenticate via the existing login API endpoint with test credentials,  and then send all subsequent requests to the existing search API with appropriate headers and cookies to work with our existing authentication. 


I'm also happy to set up a zoom meeting to discuss further if you really feel strongly that these changes should be upstreamed and really do require a new mode of authentication. Please let me know if thats what you would prefer and I can find a time for us to meet
